devel/README.md

```diff
@@ -1,7 +1,341 @@
 # Developing for Istio
 
-See also the [broker](https://github.com/istio/broker/blob/master/doc/dev/development.md) and [mixer](https://github.com/istio/mixer/blob/master/doc/dev/development.md)
-development setup and guidelines
+This document is intended to be the canonical source of truth for things like
+supported toolchain versions for building Istio components like Mixer
+(which is used throughout this document as an example when referring to a specific
+component repo or directory).
+If you find a requirement that this doc does not capture, or if you find other 
+docs with references to requirements that are not simply links to this doc,
+please [submit an issue](https://github.com/istio/istio/issues/new).
+
+This document is intended to be relative to the branch in which it is found.
+It is guaranteed that requirements will change over time for the development
+branch, but release branches should not change.
+
+- [Prerequisites](#prerequisites)
+  - [Setting up Go](#setting-up-go)
+  - [Setting up Bazel](#setting-up-bazel)
+  - [Setting up Docker](#setting-up-docker)
+  - [Setting up personal access token](#setting-up-personal-access-token)
+- [Git workflow](#git-workflow)
+  - [Fork the main repository](#fork-the-main-repository)
+  - [Clone your fork](#clone-your-fork)
+  - [Enable pre commit hook](#enable-pre-commit-hook)
+  - [Enable pre push hook](#enable-pre-push-hook)
+  - [Create a branch and make changes](#create-a-branch-and-make-changes)
+  - [Keeping your fork in sync](#keeping-your-fork-in-sync)
+  - [Committing changes to your fork](#committing-changes-to-your-fork)
+  - [Creating a pull request](#creating-a-pull-request)
+  - [Getting a code review](#getting-a-code-review)
+  - [When to retain commits and when to squash](#when-to-retain-commits-and-when-to-squash)
+- [Using the code base](#using-the-code-base)
+  - [Building the code](#building-the-code)
+  - [Cleaning outputs](#cleaning-outputs)
+  - [Running tests](#running-tests)
+  - [Getting coverage numbers](#getting-coverage-numbers)
+  - [Auto-formatting source code](#auto-formatting-source-code)
+  - [Running the linters](#running-the-linters)
+  - [Running race detection tests](#running-race-detection-tests)
+  - [Adding dependencies](#adding-dependencies)
+  - [About testing](#about-testing)
+- [Local development scripts](#collection-of-scripts-and-notes-for-developing-for-istio)
+- [MacOS tips](#macos-tips)
+
+Other docs you should look at:
+
+- Each components additional development docs like [Mixer](https://github.com/istio/mixer/tree/master/doc/dev/development.md)'s
```

Comment:
Well not every components has a specific dev guide and the list should be on
https://github.com/istio/istio/blob/master/README.md#repositories (though broker isn't...)